### ISSUE TITLE
Crash on new ingredient import (brewtarget v4.0.4) at startup time

### ISSUE THREAD

**Issue by bebenlebricolo** [Brewtarget/brewtarget#828]
When I open brewtarget 4.0.4, I can see the db migration popup (New ingredients etc are available. Would you like to add them to your database?).
I had the notification to migrate it from version 10 to 13 (if I remember correctly) ; I don't get this popup anymore.

Adding new ingredients popup :
When I click on "Yes", brewtarget ends up crashing.
If I click on "No", then I can boot up to brewtarget without modifying the database.

I gathered a crash log which seems to be a FALSE assert in TypeLookup.cpp : 
`ERROR : ASSERT: "false" in file ../brewtarget/src/utils/TypeLookup.cpp, line 89  [utils/TypeLookup.cpp:89]`

[brewtarget_4.0.4_db_crash_log.txt](https://github.com/user-attachments/files/17088674/brewtarget_4.0.4_db_crash_log.txt)

System : Arch Linux 6.10.10-arch1-1  x86_64 GNU/Linux
I'm trying to reproduce it with a build from the develop branch ; will give some news if I manage to do so.

**Comment by matty0ung** [comment]
I will take a look...

**Comment by bebenlebricolo** [comment]
Just tried with commit [15bdf1a8d16054aed3a70d99480ebbf582092e21](https://github.com/Brewtarget/brewtarget/commit/15bdf1a8d16054aed3a70d99480ebbf582092e21) and the same failure happens.
(from a fresh build).

**Comment by matty0ung** [comment]
I've managed to reproduce it locally.  Looks like there are a couple of bugs that show up when you do the recipe imports.  I've fixed one and am looking at the other.  (Broadly speaking the software now stores data more in line with the BeerJSON data model, but it still needs to be able to read and write based on the older BeerXML data model.  Mostly this is straightforward, but in a few places, we need to remember to make sensible assumptions or deductions for things that either less sophisticated or not present in BeerXML.)

**Comment by matty0ung** [comment]
OK, I have fixes for the crashes.  There's a bit more work to do to make a couple of fields the XML import work again.  I pretty much know what to do.  Hope to have a PR later this week.

**Comment by bebenlebricolo** [comment]
Thanks @matty0ung, always on the breach (it was sunday, and still you've worked on it, I hope you're not driving into burnout!)
By the way (but you might have discovered and fixed the issues as well), I noticed that when I cliked on "No" on the dialog, I could boot to brewtarget.
But when I created a new recipe, then it crashes.
When you restart brewtarget, the new recipe appears in the tree, but you cannot open it (double click) ; it crashes straight away as well.

I believe this is caused by the same issue with the db transactions.

**Comment by matty0ung** [comment]
Ah, thanks for the heads-up.  I _think_ that is a separate issue.  I'll make a new ticket for it.

**Comment by matty0ung** [comment]
The PR has fixes for the issues you raised.  I've spotted another (less severe) bug myself (https://github.com/Brewtarget/brewtarget/issues/831) but I might do a separate PR for that, as I'd like to merge the current one to test out Windows binary signing!

**Comment by bebenlebricolo** [comment]
Hmmmm I tried after the merge, and I can still reproduce the issue.
This is reproducible with the default database by the way.
1. Remove the ~/.config/brewtarget directory
2. Startup Brewtarget (warns about data migration, I've done another Issue for this #834)
3. Click on "Ok" to continue the migration 
4. Then you'll have the "Merge Database" popup
5. Click on Yes -> Crash.

Tested on commit [Merge pull request #830 from matty0ung/usability](https://github.com/Brewtarget/brewtarget/commit/765a21407634a11b73353e19449ecfed5f2dff76).

**Comment by bebenlebricolo** [comment]
Ooooooh nevermind ! 
I had a dual configuration on my machine:
* the brewtarget version installed with yay - package manager located in /usr/bin/brewtarget
* the version locally built from the repository (with the fixes) and installed in /usr/local/bin/brewtarget

The second version doesn't have the issue anymore.
So it did the trick, thanks @matty0ung  ;  no need to investigate further

**Comment by matty0ung** [comment]
No worries.  I get in a headache even switching branches.  It's one of the reasons I do development in both brewtarget and brewken.  I have both set up to compile and install locally, which means I can do small fixes directly in brewtarget, whilst simultaneously working on bigger features in brewken (which I then backport to brewtarget).

**Comment by matty0ung** [comment]
Fix included in [4.0.5 release](https://github.com/Brewtarget/brewtarget/releases/tag/v4.0.5).

**Comment by mzawodniok** [comment]
I tried v. 4.1.1 for Macos (on Apple silicon hardware) and it has segfault on clean database - new location for files. 
> zsh: segmentation fault  ./brewtarget_4.1.1_MacOS.app/Contents/MacOS/Brewtarget
[CMD-line-Add-ingredient-segfault.txt](https://github.com/user-attachments/files/20617226/CMD-line-Add-ingredient-segfault.txt)

When reopened (skipping adding new ingredients part) there is only one ingredient in hops - "Agnus". When I tried to delete it from database (in Brewtarget GUI) it ended with "bus error" ... so similar memory violation:

> zsh: bus error  ./brewtarget_4.1.1_MacOS.app/Contents/MacOS/Brewtarget
[CMD-line-Remove-Agnus-hop-buss-error.txt](https://github.com/user-attachments/files/20617223/CMD-line-Remove-Agnus-hop-buss-error.txt)

Also, when I retried the ingredients import at startup (with the one hop in it), the process moved a little further. It populated all hops and ONE malt ("Acid malt") ... and segfault-ed. 

Attachments:
[CMD-line-Remove-Agnus-hop-buss-error.txt](https://github.com/user-attachments/files/20617223/CMD-line-Remove-Agnus-hop-buss-error.txt)

[CMD-line-Add-ingredient-segfault.txt](https://github.com/user-attachments/files/20617226/CMD-line-Add-ingredient-segfault.txt)
